atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java

```diff
@@ -336,11 +342,91 @@ public void hostIsAutomaticallyAddedOnRefresh() {
-        assertThat(poolServers).containsExactlyInAnyOrder(CASS_SERVER_1, CASS_SERVER_2, CASS_SERVER_3);
+        assertThat(poolServers.keySet()).containsExactlyInAnyOrder(CASS_SERVER_1, CASS_SERVER_3);
+        assertThat(clientPool.getAbsentHostTracker().returnPool(CASS_SERVER_2)).isPresent();
+    }
+
+    @Test
+    public void throwsWhenNoServersInPoolAndServersAddedPresentButAreInvalid() {
+        setupHostsWithInconsistentTopology(ImmutableSet.of(CASS_SERVER_1));
+        setupThriftServers(ImmutableSet.of(CASS_SERVER_1.proxy()));
+        when(config.autoRefreshNodes()).thenReturn(true);
+        setCassandraServersTo(CASS_SERVER_1);
+        assertThatThrownBy(this::createClientPool).isInstanceOf(SafeIllegalStateException.class);
+    }
+
+    @Test
+    public void reuseAbsentHostsContainerIfPresent() {
+        setupThriftServers(ImmutableSet.of(CASS_SERVER_1.proxy(), CASS_SERVER_2.proxy()));
+        when(config.autoRefreshNodes()).thenReturn(true);
+
+        setCassandraServersTo(CASS_SERVER_1, CASS_SERVER_2);
+        createClientPool();
+        assertThat(poolServers.keySet()).containsExactlyInAnyOrder(CASS_SERVER_1, CASS_SERVER_2);
+        CassandraClientPoolingContainer cassandraServerTwoContainer = poolServers.get(CASS_SERVER_2);
+
+        setCassandraServersTo(CASS_SERVER_1);
+        deterministicExecutor.tick(POOL_REFRESH_INTERVAL_SECONDS, TimeUnit.SECONDS);
+        assertThat(poolServers.keySet()).containsExactlyInAnyOrder(CASS_SERVER_1);
+
+        setCassandraServersTo(CASS_SERVER_1, CASS_SERVER_2);
+        deterministicExecutor.tick(POOL_REFRESH_INTERVAL_SECONDS, TimeUnit.SECONDS);
+        assertThat(poolServers.keySet()).containsExactlyInAnyOrder(CASS_SERVER_1, CASS_SERVER_2);
+        assertThat(cassandraServerTwoContainer).isEqualTo(poolServers.get(CASS_SERVER_2));
+    }
+
+    @Test
+    public void reuseContainerIfPreviouslyInvalid() {
+        setupThriftServers(ImmutableSet.of(CASS_SERVER_1.proxy(), CASS_SERVER_2.proxy()));
+        when(config.autoRefreshNodes()).thenReturn(true);
+
+        setCassandraServersTo(CASS_SERVER_1);
+        CassandraClientPoolImpl cassandraClientPool = createClientPool();
+        CassandraClientPoolingContainer cassandraServerTwoContainer = mock(CassandraClientPoolingContainer.class);
+        assertThat(poolServers.keySet()).containsExactlyInAnyOrder(CASS_SERVER_1);
+        cassandraClientPool
+                .getAbsentHostTracker()
+                .trackAbsentCassandraServer(CASS_SERVER_2, cassandraServerTwoContainer);
+
+        setCassandraServersTo(CASS_SERVER_1, CASS_SERVER_2);
+        deterministicExecutor.tick(POOL_REFRESH_INTERVAL_SECONDS, TimeUnit.SECONDS);
+        assertThat(poolServers.keySet()).containsExactlyInAnyOrder(CASS_SERVER_1, CASS_SERVER_2);
+        assertThat(cassandraServerTwoContainer).isEqualTo(poolServers.get(CASS_SERVER_2));
+    }
+
+    @Test
+    public void refreshHandlesNothingToAddOrRemove() {
```

Comment:
what is the expected state for "handling"? Perhaps `noServersAddedOrRemovedWhenCassandraServerListStatic` or something like that.